tests/dgram-speed-client.cpp

```diff
@@ -0,0 +1,229 @@
+    cli.add_option("-c,--servercert", server_cert, "Path to server certificate to use")
+            ->type_name("FILE")
+            ->capture_default_str()
+            ->check(CLI::ExistingFile);
+
+    uint64_t size = 1'000'000'000;
+    cli.add_option("-S,--size", size, "Amount of data to transfer.");
+
+    bool pregenerate = true;
+    cli.add_flag("-g,--pregenerate", pregenerate, "Pregenerate all stream data to send into RAM before starting");
+
+    size_t dgram_size = 0;
+    cli.add_option("--dgram-size", dgram_size, "Datagram size to send");
+
+    // TODO: make this optional
+    std::string cert{"./clientcert.pem"}, key{"./clientkey.pem"};
+    cli.add_option("-C,--certificate", key, "Path to client certificate for client authentication")
+            ->type_name("FILE")
+            ->capture_default_str()
+            ->check(CLI::ExistingFile);
+    cli.add_option("-K,--key", key, "Path to client key to use for client authentication")
+            ->type_name("FILE")
+            ->capture_default_str()
+            ->check(CLI::ExistingFile);
+
+    try
+    {
+        cli.parse(argc, argv);
+    }
+    catch (const CLI::ParseError& e)
+    {
+        return cli.exit(e);
+    }
+
+    struct send_data
+    {
+        std::shared_ptr<Stream> stream;
+        std::atomic<bool> active = false;
+        std::basic_string<uint8_t> msg{};
+        uint64_t size;
+        uint64_t dgram_size;
+        uint64_t n_iter;
+        std::atomic<bool> is_sending = false;
+        std::atomic<bool> is_done = false;
+        std::promise<void> run_prom;
+        std::future<void> running = run_prom.get_future();
+        std::atomic<bool> failed = false;
+
+        std::basic_string<std::byte> hash;
+        uint8_t checksum = 0;
+        gnutls_hash_hd_t sent_hasher, recv_hasher;
+
+        send_data() {}
+        send_data(uint64_t _total_size, uint64_t _dgram_size) : size{_total_size}, dgram_size{_dgram_size}
+        {
+            n_iter = size / dgram_size + 1;
+
+            log::warning(test_cat, "Preparing to send {} datagrams of max size {}", n_iter, size);
+
+            uint8_t v{0};
+
+            while (msg.size() < dgram_size)
+                msg += v++;
+
+            gnutls_hash_init(&sent_hasher, GNUTLS_DIG_SHA3_256);
+            gnutls_hash_init(&recv_hasher, GNUTLS_DIG_SHA3_256);
+        }
+
+        ~send_data()
+        {
+            gnutls_hash_deinit(sent_hasher, nullptr);
+            gnutls_hash_deinit(recv_hasher, nullptr);
+        }
+    };
+
+    setup_logging(log_file, log_level);
+
+    Network client_net{};
+
+    auto client_tls = GNUTLSCreds::make(key, cert, server_cert);
+
+    send_data* d_ptr;
+
+    stream_close_callback stream_closed = [&](Stream& s, uint64_t errcode) {
+        size_t i = s.stream_id() >> 2;
+        log::critical(test_cat, "Stream {} (rawid={}) closed (error={})", i, s.stream_id(), errcode);
+    };
+
+    dgram_data_callback recv_dgram_cb = [&](bstring data) {
```

Comment:
Looking at this implementation, it seems that we have a couple issues with the argument passed here:
- when we don't have a split packet (because either we don't have packet splitting, or we receive a small packet) we are still allocating a `bstring` that we could avoid.
- there's no way to tell *which* connection sent me this datagram.  I was looking to make this server persistent, but right now I can't which is going to be a problem when this is used in a many-to-one client-to-server architecture.

The second issue is easy to fix (just pass the connection_interface along), but the first issue is trickier because *sometimes* we want to pass a `bstring` that we already have on hand (i.e. from reassembling a split packet), but other times we we don't.

Reply:
Resolving this per off-github discussion, to be addressed in #29 ongoing